### PR TITLE
Fix compiler crash in supertypes() and get_field()

### PR DESCRIPTION
**Pull request by ark3**

### CODE REVIEW

File: quarkc/compiler.py [hunk 1]
```diff
@@ -350,6 +350,10 @@ def supertypes(self):
         for sup in self.supertypes(self.type):
             yield sup
 
+    @overload(AST)
```
Comment (itamarst) [inline quarkc/compiler.py]: What's "supertypes" do? And why does it sometimes take an argument and sometimes not? (Some docstrings might help).

File: quarkc/helpers.py [hunk 2]
```diff
@@ -38,6 +38,10 @@ def __init__(self, loc, warning):
     def __str__(self):
         return "%s: warning: %s" % (lineinfo(self.loc),  self.warning)
 
+@dispatch(AST, object)
```
Comment (itamarst) [inline quarkc/helpers.py]: Docstring for `get_field` in general would be helpful, and explanation for what this particular variant does.

File: quarkc/test/compile/bad-method-expr.err [hunk 3]
```diff
@@ -0,0 +1 @@
+bad-method-expr.q:4:24:type mismatch: expected 1.quark.String, got 0.bad_method_expr.Foo.bar
```
Comment (itamarst) [inline quarkc/test/compile/bad-method-expr.err]: On the one hand that's a legit error. ON the other hand... if we don't support references to functions or methods, shouldn't it fail do to that? Also where did `Foo` come from?

Comment (itamarst) [inline quarkc/test/compile/bad-method-expr.err]: (If `Foo` is a typo, is this `.err` actually being used?)

**Comment by itamarst** [comment]
@ark3 a bunch of questions/notes above.

**Comment by ark3** [comment]
My problem with adding doc strings is that I don't know what should go there. I made a minimal set of changes to address the crashes. I'll try my best to answer the questions, but adding incorrect doc strings would probably be bad...

> What's "supertypes" do? And why does it sometimes take an argument and sometimes not? (Some docstrings might help).

I think it returns set of types associated with the element passed in, or if nothing is passed in, then the set of types associated with `self.type`.

> Docstring for get_field in general would be helpful, and explanation for what this particular variant does.

This seems to return the definition AST object associated with a field. This variant, like the above variant for supertypes, is the catch-all that returns a result that will provoke a compiler error.

> On the one hand that's a legit error. ON the other hand... if we don't support references to functions or methods, shouldn't it fail do to that? Also where did Foo come from?

Those references are not first-class objects in Quark, but they do exist conceptually. Are you suggesting that we should have a very different error message? `Foo` is the name of the class.

**Comment by itamarst** [comment]
Ah, I misread where `Foo` came from, never mind that.

**Comment by itamarst** [comment]
1. Maybe for e.g. that `get_field` you can say something like "Catch-all `get_field` to ensure <whatever it ensures by having default response>"? Given the code structure it's hard to tell which code calls what, so it's good to have some explanation for each variant.
2. My point re error message that the error message isn't covering the underlying problem, e.g. you can imagine invalid code like `SomeClass.someMethod;`. But if it's hard to fix the error message it's not very important to deal with this.

**Pull request merged**